PORTAL/jobs/_utils.py

```diff
@@ -256,67 +310,83 @@ def _normalize_columns(cls, specs, names, maxwidth):
         #        ('started,created', 'start / (created)', 21, None),
         #    )
 
-        return columns, names
+        return columns, names_list
 
     @property
-    def colnames(self):
+    def colnames(self) -> List[str]:
         try:
-            return self._colnames
+            return self._colnames  # type: ignore[has-type]
         except AttributeError:
             self._colnames = [c.name for c in self.columns]
             return self._colnames
 
-    def render(self, rows, *, numrows=None, total=None):
+    def render(
+            self,
+            rows: Iterable["TableRow"],
+            *,
+            numrows: Optional[int] = None,
+            total: Optional[int] = None
+    ) -> Iterable[str]:
         # We use a default format here.
         header = [self.div, self.header, self.div]
-        rows = self._render_rows(rows)
-        rows = RenderingRows(rows, header, numrows)
+        rendered_rows = self._render_rows(rows)
+        rendering_rows = RenderingRows(rendered_rows, header, numrows)
         yield from header
-        yield from rows
-        yield div
-        yield from rows.render_count(total)
-
-    def render_rows(self, rows, periodic=None, numrows=None):
-        rows = self._render_rows(rows)
-        return RenderingRows(rows, periodic, numrows)
-
-    def _render_rows(self, rows):
+        yield from rendering_rows
+        yield self.div
+        yield from rendering_rows.render_count(total)
+
+    def render_rows(
+            self,
+            rows: Iterable["TableRow"],
+            periodic: Optional[Iterable[str]] = None,
+            numrows: Optional[int] = None
+    ) -> "RenderingRows":
+        rendered_rows = self._render_rows(rows)
+        return RenderingRows(rendered_rows, periodic, numrows)
+
+    def _render_rows(self, rows: Iterable["TableRow"]) -> Iterable[str]:
         fmt = self.rowfmt
         for row in rows:
             values = row.render_values(self.colnames)
             yield fmt.format(*values)
 
-    def _render_row(self, row):
+    def _render_row(self, row: "TableRow") -> str:
         values = row.render_values(self.colnames)
         return self.rowfmt.format(*values)
 
 
 class RenderingRows:
 
-    def __init__(self, rows, periodic=None, numrows=None):
+    def __init__(
+            self,
+            rows: Iterable[str],
+            periodic: Optional[Any] = None,
+            numrows: Optional[int] = None
+    ):
         if not periodic:
             periodic = None
         elif isinstance(periodic, str):
             periodic = [periodic]
         if not numrows:
             try:
-                numrows = len(rows)
+                numrows = len(list(rows))
             except TypeError:
                 numrows = None
         self.rows = iter(rows)
         self.periodic = periodic
         self.numrows = numrows
 
         self.count = 0
-        self.pending = []
+        self.pending: List[str] = []
```

Comment:
This use of annotation keeps weirding me out.  It might be more clear to move the annotation to class-level:

```Python
class RenderingRows:

    pending: List[str]
    ...
```